### PR TITLE
Extend head tracker so it tracks all forks

### PR DESCRIPTION
**Pull request by samsondav**
- OnNewHead renamed to OnNewLongestChain and now receives a head along with its parents up to N blocks deep
- This is required for re-org detection and more efficient confirmation
ensurance in the bulletproof tx manager
- We now use idempotent insertion of new heads. A side effect of this is that the head tracker can be safely run in parallel across multiple nodes, moving us closer to our horizontal scaling OKR.

### CODE REVIEW

File: core/services/head_tracker_test.go [hunk 1]
```diff
@@ -20,40 +20,55 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func firstHead(t *testing.T, store *strpkg.Store) models.Head {
+	h := models.Head{}
+	if err := store.GetRawDB().Order("number asc").First(&h).Error; err != nil {
```
Comment (se3000) [inline core/services/head_tracker_test.go:25]: Do we need GetRawDB here? Why do we need to skip the lock?

File: core/store/orm/orm.go [hunk 2]
```diff
@@ -1315,6 +1354,11 @@ func (ct Connection) initializeDatabase() (*gorm.DB, error) {
 // NOTE: Now we no longer support SQLite, perhaps this can be tuned?
 const BatchSize = 100
 
+// NOTE: Skips advisory lock, use caution!
+func (orm *ORM) GetRawDB() *gorm.DB {
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/store/orm/orm.go:1358]: Good question. Couple of reasons:

1. I use it extensively in the TxManager PR upcoming and I pulled this code right out of there.
2. I want to start moving us away from the *Big Global Lock* (tm) as part of the horizontal scaling OKR. Many functions that access the DB through a lock do not actually need to do this.

File: core/services/head_tracker_test.go (hunk 1 above)
Comment (samsondav) [reply core/services/head_tracker_test.go:25]: Why would we need a global lock for this query?

File: core/services/head_tracker.go [hunk 3]
```diff
@@ -210,30 +202,44 @@ func (ht *HeadTracker) subscribe() bool {
...
-				switch err.(type) {
-				case errBlockNotLater:
-					logger.Warn(err)
-				default:
-					logger.Error(err)
+				return err
+			}
+			if prevHead == nil || head.Number > prevHead.Number {
+				headWithChain, err := ht.store.Chain(head.Hash, chainDepth)
+				if err != nil {
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/head_tracker.go:228]: Agree that it's complicated, I'll clean it up.

File: core/services/job_subscriber.go [hunk 4]
```diff
@@ -35,20 +35,20 @@ type JobSubscriber interface {
...
-	jobsMutex                 *sync.RWMutex
-	runManager                RunManager
-	jobResumer                SleeperTask
-	resumeRunsOnNewHeadWorker *resumeRunsOnNewHeadWorker
+	store                             *store.Store
+	jobSubscriptions                  map[string]JobSubscription
+	jobsMutex                         *sync.RWMutex
+	runManager                        RunManager
+	jobResumer                        SleeperTask
+	resumeRunsOnNewLongestChainWorker *resumeRunsOnNewLongestChainWorker
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/job_subscriber.go:43]: Yeah this is a bit of a rubyism... let me rename it.

**Pull request merged**